lib/utils/utils.go

```diff
@@ -140,6 +142,135 @@ func ToJSON(data []byte) ([]byte, error) {
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// ignore root directory
+		if path == srcDir {
+			return nil
+		}
+
+		if fi.IsDir() {
+			// create directory for the target file
+			targetDir := filepath.Join(dstDir, strings.TrimPrefix(path, srcDir))
+			err = os.MkdirAll(targetDir, constants.SharedDirMask)
+			if err != nil {
+				return trace.ConvertSystemError(err)
+			}
+			// copy sub-directories recursively
+			return nil
+		}
+
+		relativePath := strings.TrimPrefix(filepath.Dir(path), srcDir)
+		targetDir := filepath.Join(dstDir, relativePath)
+
+		// copy file, preserve permissions
+		toFileName := filepath.Join(targetDir, filepath.Base(fi.Name()))
+		err = CopyFileWithPerms(toFileName, path, fi.Mode())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
+	})
+	return trace.Wrap(err)
+}
+
+// CopyFileWithPerms copies the contents from src to dst atomically.
+// Uses CopyReaderWithPerms for its implementation - see function documentation
+// for details of operation
+func CopyFileWithPerms(dst, src string, perm os.FileMode) error {
+	in, err := os.Open(src)
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	defer in.Close()
+	return CopyReaderWithPerms(dst, in, perm)
+}
+
+// CopyReaderWithPerms copies the contents from src to dst atomically.
+// If dst does not exist, CopyReaderWithPerms creates it with permissions perm.
+// If the copy fails, CopyReaderWithPerms aborts and dst is preserved.
+// Adopted with modifications from https://go-review.googlesource.com/#/c/1591/9/src/io/ioutil/ioutil.go
+func CopyReaderWithPerms(dst string, src io.Reader, perm os.FileMode) error {
+	return CopyReaderWithOptions(dst, src, PermOption(perm))
+}
+
+// CopyReaderWithOptions copies the contents from src to dst atomically.
+// If dst does not exist, CopyReaderWithOptions creates it.
+// Callers choose the options to apply on the resulting file with options
+func CopyReaderWithOptions(dst string, src io.Reader, options ...FileOption) error {
+	tmp, err := ioutil.TempFile(filepath.Dir(dst), "")
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	cleanup := func() {
+		err := os.Remove(tmp.Name())
+		if err != nil {
+			log.WithError(err).Warnf("Failed to remove %q.", tmp.Name())
+		}
+	}
+
+	_, err = io.Copy(tmp, src)
+	if err != nil {
+		tmp.Close()
+		cleanup()
+		return trace.ConvertSystemError(err)
+	}
+	if err = tmp.Close(); err != nil {
+		cleanup()
+		return trace.ConvertSystemError(err)
+	}
+	for _, option := range options {
+		if err = option(tmp.Name()); err != nil {
+			cleanup()
+			return trace.ConvertSystemError(err)
+		}
+	}
+	err = os.Rename(tmp.Name(), dst)
+	if err != nil {
+		cleanup()
+		return trace.ConvertSystemError(err)
+	}
+	return nil
+}
+
+// FileOption defines a functional option to apply to specified path
+type FileOption func(path string) error
+
+// PermOption changes the file permissions on the specified file
+// to perm
+func PermOption(perm os.FileMode) FileOption {
```

Comment:
Nit: Not that it matters much, but I think using the word "Option" in the option is a bit redundant. I would just call them `WithFileMode()`, `WithOwner()`.